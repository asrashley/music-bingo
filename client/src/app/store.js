import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import { connectRouter } from 'connected-react-router';
import { createBrowserHistory } from 'history';
import { reducer as formReducer } from 'redux-form';

import cardsReducer from '../cards/cardsSlice';
import gamesReducer from '../games/gamesSlice';
import ticketsReducer from '../tickets/ticketsSlice';
import userReducer from '../user/userSlice';
import adminReducer from '../admin/adminSlice';
import usersMiddleware from '../user/userMiddleware';

export const history = createBrowserHistory();

export const store = configureStore({
  middleware: [...getDefaultMiddleware(), usersMiddleware, logger],
  reducer: {
    admin: adminReducer,
    cards: cardsReducer,
    form: formReducer,
    games: gamesReducer,
    tickets: ticketsReducer,
    router: connectRouter(history),
    user: userReducer,
  },
});
