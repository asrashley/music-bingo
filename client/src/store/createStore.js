import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import { connectRouter } from 'connected-react-router';

import adminReducer from '../admin/adminSlice';
import directoriesReducer from '../directories/directoriesSlice';
import gamesReducer from '../games/gamesSlice';
import messagesReducer from '../messages/messagesSlice';
import settingsReducer from '../settings/settingsSlice';
import systemReducer from '../system/systemSlice';
import ticketsReducer from '../tickets/ticketsSlice';
import userReducer from '../user/userSlice';
import usersMiddleware from '../user/userMiddleware';
import messagesMiddleware from '../messages/messagesMiddleware';

import { history } from './history';

export function createStore(preloadedState, logging) {
  const middleware = [...getDefaultMiddleware(), usersMiddleware, messagesMiddleware];
  if (logging === undefined) {
    logging = (process.env.NODE_ENV === 'development');
  }
  if (logging) {
    middleware.push(logger);
  }
  return configureStore({
    middleware,
    reducer: {
      admin: adminReducer,
      directories: directoriesReducer,
      games: gamesReducer,
      messages: messagesReducer,
      settings: settingsReducer,
      system: systemReducer,
      tickets: ticketsReducer,
      router: connectRouter(history),
      user: userReducer,
    },
    preloadedState
  });
}
