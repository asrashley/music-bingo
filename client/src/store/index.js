export { createStore } from './createStore';
export { initialState } from './initialState';
export { history } from './history';
