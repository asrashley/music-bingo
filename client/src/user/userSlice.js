import { isEmail } from 'validator';
import { createSlice } from '@reduxjs/toolkit';
import { api } from '../endpoints';

export const userChangeListeners = {};

export const userSlice = createSlice({
  name: 'user',
  initialState: {
    username: '',
    email: '',
    pk: -1,
    options: {
      colourScheme: 'blue',
      maxTickets: 2,
      columns: 5,
      rows: 3,
    },
    isFetching: false,
    registering: false,
    didInvalidate: false,
    error: null,
    lastUpdated: null,
    activeGame: null,
    groups: {
    },
    accessToken: localStorage.getItem('accessToken'),
    refreshToken: localStorage.getItem('refreshToken'),
    tokenFetching: false,
    newUser: {
      isChecking: false,
      valid: false,
    },
  },
  reducers: {
    requestUser: (state, action) => {
      state.isFetching = true;
    },
    requestLogin: (state, action) => {
      const { body } = action.payload;
      const { username } = body;
      state.isFetching = true;
      state.username = username;
    },
    failedLoginUser: (state, action) => {
      const { body, status, error, timestamp } = action.payload;
      if (status === 401){
        const { username } = body;
        if (isEmail(username)) {
          state.error = 'Email address or password is incorrect';
        } else {
          state.error = 'Username or password is incorrect';
        }
      } else {
        state.error = error;
      }
      state.lastUpdated = timestamp;
      state.isFetching = false;
    },
    confirmLogoutUser: state => {
      state.isFetching = false;
      state.pk = -1;
      state.username = '';
      state.error = null;
      state.didInvalidate = true;
      state.accessToken = null;
      state.refreshToken = null;
    },
    registeringUser: (state, action) => {
      const { body } = action.payload;
      const { username, email } = body;
      state.registering = true;
      state.email = email;
      state.username = username;
    },
    failedRegisterUser: (state, action) => {
      const { error, timestamp } = action.payload;
      state.registering = false;
      state.error = error;
      state.lastUpdated = timestamp;
    },
    userRegistered: (state, action) => {
      const { payload, timestamp } = action.payload;
      const { user, accessToken, refreshToken } = payload;
      for (let key in user) {
        const value = user[key];
        if (key === 'timestamp' || value === undefined) {
          continue;
        }
        state[key] = value;
      }
      state.groups = {};
      user.groups.forEach(g => state.groups[g] = true);
      state.isFetching = false;
      state.registering = false;
      state.error = null;
      state.lastUpdated = timestamp;
      state.didInvalidate = false;
      if (accessToken) {
        localStorage.setItem('accessToken', accessToken);
      }
      if (refreshToken) {
        localStorage.setItem('refreshToken', refreshToken);
      }
    },
    receiveUser: (state, action) => {
      const { timestamp, payload } = action.payload;
      for (let key in payload) {
        const value = payload[key];
        if (key === 'timestamp' || value === undefined) {
          continue;
        }
        state[key] = value;
      }
      state.groups = {};
      payload.groups.forEach(g => state.groups[g] = true);
      state.isFetching = false;
      state.registering = false;
      state.error = null;
      state.lastUpdated = timestamp;
      state.didInvalidate = false;
      if (payload.accessToken) {
        localStorage.setItem('accessToken', payload.accessToken);
      }
      if (payload.refreshToken) {
        localStorage.setItem('refreshToken', payload.refreshToken);
      }
    },
    failedFetchUser: (state, action) => {
      const { timestamp, error } = action.payload;
      state.isFetching = false;
      state.registering = false;
      state.lastUpdated = timestamp;
      if (state.pk > 0){
        state.error = error;
      }
      state.didInvalidate = true;
    },
    requestCheckNewUser: (state, action) => {
      const { body } = action.payload;
      const {username, email} = body;
      state.newUser = {
        username,
        email,
        isChecking: true,
        isValid: null,
      };
    },
    failedCheckNewUser: (state, action) => {
      state.newUser.isChecking = false;
      state.newUser.valid = false;
    },
    successCheckNewUser: (state, action) => {
      state.newUser.isChecking = false;
      state.newUser.valid = true;
    },
    failedResetUser: (state, action) => {
      const { error } = action.payload;
      state.isFetching = false;
      state.error = error;
    },
    confirmPasswordReset: (state, action) => {
    },
    setActiveGame: (state, action) => {
      let gamePk = action.payload;
      if (typeof (gamePk) === "string") {
        gamePk = parseInt(gamePk);
      }
      if (typeof (gamePk) === "number" && !isNaN(gamePk)) {
        state.activeGame = gamePk;
      }
    },
    requestToken: (state) => {
      state.tokenFetching = true;
    },
    failedFetchToken: (state, action) => {
      const { timestamp, error } = action.payload;
      state.tokenFetching = false;
      state.error = error;
      state.lastUpdated = timestamp;
      state.accessToken = null;
    },
    receiveToken: (state, action) => {
      const { timestamp, payload } = action.payload;
      state.tokenFetching = false;
      state.accessToken = payload.accessToken;
      state.lastUpdated = timestamp;
    },
  },
});

export const { setActiveGame } = userSlice.actions;

function fetchUser() {
  const success = [
    userSlice.actions.receiveUser,
    ...Object.values(userChangeListeners),
  ];
  return api.getUserInfo({
    rejectErrors: false,
    before: userSlice.actions.requestUser,
    failure: userSlice.actions.failedFetchUser,
    success,
  });
}

function shouldFetchUser(state) {
  const { user } = state;
  if (user.pk <= 0) {
    return true;
  }
  if (user.isFetching) {
    return false;
  }
  return user.didInvalidate;
}

export function fetchUserIfNeeded() {
  return (dispatch, getState) => {
    if (shouldFetchUser(getState())) {
      return dispatch(fetchUser());
    }
    return Promise.resolve();
  };
}

export function checkUser({ username, email }) {
  return api.checkUser({
    body: { username, email },
    before: userSlice.actions.requestCheckNewUser,
    failure: userSlice.actions.failedCheckNewUser,
    success: userSlice.actions.successCheckNewUser,
  });
}

export function loginUser(user) {
  return api.login({
    body: user,
    noAccessToken: true,
    before: userSlice.actions.requestLogin,
    failure: userSlice.actions.failedLoginUser,
    success: userSlice.actions.receiveUser
  });
}

export function logoutUser() {
  localStorage.removeItem("accessToken");
  localStorage.removeItem("refreshToken");
  return api.logout({
    success: userSlice.actions.confirmLogoutUser,
    failure: userSlice.actions.failedLogout
  });
}

export function registerUser(user) {
  return api.registerUser({
    body: user,
    noAccessToken: true,
    rejectErrors: false,
    before: userSlice.actions.registeringUser,
    success: userSlice.actions.userRegistered,
    failure: userSlice.actions.failedRegisterUser
  });
}

export function passwordResetUser(user) {
  return api.passwordReset({
    body: user,
    noAccessToken: true,
    rejectErrors: false,
    success: userSlice.actions.confirmPasswordReset,
    failure: userSlice.actions.failedResetUser
  });
}

export function refreshAccessToken(refreshToken) {
  return api.refreshToken({
    refreshToken,
    before: userSlice.actions.requestToken,
    success: userSlice.actions.receiveToken,
    failure: userSlice.actions.failedFetchToken,
  });
}

export const initialState = userSlice.initialState;

export default userSlice.reducer;

api.actions.refreshAccessToken = refreshAccessToken;
