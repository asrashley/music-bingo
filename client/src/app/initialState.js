import { initialState as cards } from '../cards/cardsSlice';
import { initialState as games } from '../games/gamesSlice';
import { initialState as tickets } from '../tickets/ticketsSlice';
import { initialState as user } from '../user/userSlice';

export const initialState = Object.freeze({
    cards,
    games,
    tickets,
    user,
});
