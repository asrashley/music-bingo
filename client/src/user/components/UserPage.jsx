import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { reverse } from 'named-urls';

/* components */
import { AdminGameActions } from '../../games/components/AdminGameActions';

/* actions */
import { fetchUserIfNeeded } from '../../user/userSlice';

/* selectors */
import { getUser } from '../../user/userSelectors';
import { getDatabaseImportState } from '../../admin/adminSelectors';
import { getGameImportState } from '../../games/gamesSelectors';

/* data */
import routes from '../../routes';
import { initialState } from '../../app/initialState';

import '../styles/user.scss';

class UserPage extends AdminGameActions {
  static propTypes = {
    dispatch: PropTypes.func.isRequired,
    history: PropTypes.object.isRequired,
    user: PropTypes.object.isRequired,
  };

  componentDidMount() {
    const { dispatch } = this.props;
    dispatch(fetchUserIfNeeded());
  }

  render() {
    const { user, databaseImporting, gameImporting } = this.props;
    const { ActiveDialog, dialogData, error, importType } = this.state;
    let importing = {};
    let manage = '';
    if (importType === AdminGameActions.DATABASE_IMPORT) {
      importing = databaseImporting;
    } else if (importType === AdminGameActions.GAME_IMPORT) {
      importing = gameImporting;
    }
    if (user.groups.admin === true) {
      manage = (
        <React.Fragment>
          <Link className="btn btn-lg btn-primary mb-4"
            to={reverse(`${routes.listUsers}`)}>Modify Users
          </Link>
          <Link className="btn btn-lg btn-primary mb-5"
            to={reverse(`${routes.guestLinks}`)}>Guest links
          </Link>
          <button className="btn btn-lg btn-primary mb-4"
            onClick={this.onClickImportGame}>Import Game
          </button>
          <button className="btn btn-lg btn-primary mb-4"
            onClick={this.onClickImportDatabase}>Import Database
          </button>
          <button className="btn btn-lg btn-primary mb-5"
            onClick={this.exportDatabase}>Export Database
          </button>
        </React.Fragment>
      );
    }

    return (
      <div id="user-page">
        {error && <div className="alert alert-warning"
          role="alert"><span className="error-message">{error}</span></div>}
        <div className="user-details border border-secondary rounded">
          <div className="form-group row">
            <label htmlFor="username" className="col-sm-2 col-form-label field">Username</label>
            <div className="col-sm-10">
              <div className="form-control-plaintext value" id="username">
                {user.username}
              </div>
            </div>
          </div>
          <div className="form-group row">
            <label htmlFor="email" className="col-sm-2 col-form-label">Email</label>
            <div className="col-sm-10">
              <div className="form-control-plaintext" id="email">
                {user.email}
              </div>
            </div>
          </div>
        </div>
        <div className="user-commands">
          {manage}
          {user.guest.loggedIn !== true && <Link to={reverse(`${routes.changeUser}`)}
            className="btn btn-lg btn-warning change-user mt-3 mb-5">
            Change password or email address</Link>}
          <Link to={reverse(`${routes.logout}`)}
            className="btn btn-lg btn-primary logout mb-5">Log out</Link>
        </div>
        {ActiveDialog && <ActiveDialog backdrop {...dialogData} {...importing} />}
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  state = state || initialState;
  return {
    user: getUser(state, props),
    databaseImporting: getDatabaseImportState(state),
    gameImporting: getGameImportState(state),
  };
};

UserPage = connect(mapStateToProps)(UserPage);

export { UserPage };
