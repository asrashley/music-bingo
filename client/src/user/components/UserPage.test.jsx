import React from 'react';
import { act } from '@testing-library/react';
import log from 'loglevel';

import { createStore } from '../../store/createStore';
import { initialState } from '../../store/initialState';
import { fetchMock, renderWithProviders, installFetchMocks } from '../../../tests';
import { UserPage, UserPageComponent } from './UserPage';
import { userSlice } from '../userSlice';
import * as user from '../../../tests/fixtures/userState.json';

describe('UserPage component', () => {
  let apiMocks;

  beforeAll(() => {
    vi.useFakeTimers('modern');
    vi.setSystemTime(new Date('08 Feb 2023 10:12:00 GMT').getTime());
  });

  beforeEach(() => {
    apiMocks = installFetchMocks(fetchMock, { loggedIn: true });
  });

  afterEach(() => {
    fetchMock.mockReset();
    vi.clearAllMocks();
    log.resetLevel();
    apiMocks = null;
    log.resetLevel();
  });

  afterAll(() => {
    vi.useRealTimers();
    vi.restoreAllMocks();
  });

  it('renders without throwing an exception with initial state', () => {
    const store = createStore(initialState);
    const props = {
      dispatch: store.dispatch,
      history: {
        push: vi.fn()
      },
      user: initialState.user
    };
    const { getByText } = renderWithProviders(<UserPageComponent {...props} />);
    getByText('Log out');
  });

  it('updates settings once user has logged in', async () => {
    fetchMock.mockReset();
    apiMocks = installFetchMocks(fetchMock, { loggedIn: false });
    const username = 'my.username';
    const password = 'password!';
    const store = createStore(initialState);
    apiMocks.addUser({
      username,
      password,
      groups: ['users'],
      pk: 12,
    });
    const { getByText, findByText } = renderWithProviders(<UserPage />, { store });
    getByText('Log out');

    await act(() => {
      store.dispatch(userSlice.actions.receiveUser({
        timestamp: 123,
        payload: {
          username: 'my.username',
          email: 'a.user',
          groups: ['users'],
          pk: 12,
          options: user.options,
        },
      }));
    });

    await findByText('my.username');
  });

  it('renders admin user actions', async () => {
    const store = createStore({
      ...initialState,
      user
    });
    const { asFragment, findByText } = renderWithProviders(<UserPage />, { store });
    const actions = [
      'Modify Users',
      'Modify Settings',
      'Guest links',
      'Import Game',
      'Import Database',
      'Export Database',
      'Change password or email address',
      'Log out'
    ];
    await Promise.all(actions.map(item => findByText(item)));
    expect(asFragment()).toMatchSnapshot();
  });
});
