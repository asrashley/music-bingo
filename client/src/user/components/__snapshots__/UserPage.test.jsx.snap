// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserPage component renders admin user actions 1`] = `
<DocumentFragment>
  <div
    id="user-page"
  >
    <div
      class="user-details border border-secondary rounded"
    >
      <div
        class="form-group row"
      >
        <label
          class="col-sm-2 col-form-label field"
          for="username"
        >
          Username
        </label>
        <div
          class="col-sm-10"
        >
          <div
            class="form-control-plaintext value"
            id="username"
          >
            admin
          </div>
        </div>
      </div>
      <div
        class="form-group row"
      >
        <label
          class="col-sm-2 col-form-label"
          for="email"
        >
          Email
        </label>
        <div
          class="col-sm-10"
        >
          <div
            class="form-control-plaintext"
            id="email"
          >
            admin@music.bingo
          </div>
        </div>
      </div>
    </div>
    <div
      class="user-commands"
    >
      <div
        class="user-admin-actions"
      >
        <a
          class="btn btn-lg btn-primary mb-4"
          href="/user/users"
        >
          Modify Users
        </a>
        <a
          class="btn btn-lg btn-primary mb-4"
          href="/user/settings"
        >
          Modify Settings
        </a>
        <a
          class="btn btn-lg btn-primary mb-5"
          href="/user/guests"
        >
          Guest links
        </a>
        <button
          class="btn btn-primary btn-lg mb-4"
        >
          Import Game
        </button>
        <button
          class="btn btn-primary btn-lg mb-4"
        >
          Import Database
        </button>
        <button
          class="btn btn-primary btn-lg mb-4"
        >
          Export Database
        </button>
      </div>
      <a
        class="btn btn-lg btn-warning change-user mt-3 mb-5"
        href="/user/modify"
      >
        Change password or email address
      </a>
      <a
        class="btn btn-lg btn-primary logout mb-5"
        href="/user/logout"
      >
        Log out
      </a>
    </div>
  </div>
</DocumentFragment>
`;
